File: lib/pelican_installer/screens/webserver.py
"""Webserver selection screen."""

from __future__ import annotations

from textual import on
from textual.app import ComposeResult
from textual.containers import Container
from textual.screen import Screen
from textual.widgets import Button, Static

from pelican_installer.components.menu import InstallerMenu
from pelican_installer.utils.state import InstallState


class WebserverScreen(Screen[str]):
    """Screen for selecting webserver (Nginx/Apache/Caddy)."""

    CSS = """
    WebserverScreen {
        align: center middle;
    }
    """

    def __init__(self, state: InstallState) -> None:
        super().__init__()
        self.state = state

    def compose(self) -> ComposeResult:
        with Container(id="root"):
            with Container(id="card"):
                yield Static("Webserver Configuration", id="title")
                yield Static("Select your preferred webserver:", id="subtitle")
                yield InstallerMenu(id="webserver-menu")
                yield Static(
                    "Use ↑/↓, 1-3, Enter or click to select",
                    id="hint",
                )
                with Container(id="footer"):
                    yield Button("Back (b)", id="back")
                    yield Button("Close (c)", id="close")

    def on_mount(self) -> None:
        """Set up webserver options."""
        menu = self.query_one("#webserver-menu", InstallerMenu)
        menu.clear_options()
        menu.add_options(
            [
                "1) Nginx (recommended)",
                "2) Apache",
                "3) Caddy",
            ]
        )
        menu.highlighted = 0

    @on(InstallerMenu.OptionSelected)
    def handle_selection(self, event: InstallerMenu.OptionSelected) -> None:
        """Handle webserver selection."""
        if event.option_index == 0:
            self.state.webserver = "nginx"
        elif event.option_index == 1:
            self.state.webserver = "apache"
        elif event.option_index == 2:
            self.state.webserver = "caddy"

        self.dismiss("protocol")

    @on(Button.Pressed, "#back")
    def back_pressed(self) -> None:
        """Go back to main menu."""
        self.dismiss("back")

    @on(Button.Pressed, "#close")
    def close_pressed(self) -> None:
        """Handle close button."""
        self.app.exit()

    def action_request_close(self) -> None:
        """Global close action (c key)."""
        self.app.exit()

File: src/screens/webserver.rs
use crate::screens::{Screen, ScreenResult};
use crate::ui::components::SelectMenu;
use crate::utils::state::{InstallState, WebserverType};
use ratatui::prelude::*;
use ratatui::widgets::Paragraph;
use crossterm::event::KeyEvent;

pub struct WebserverScreen {
    menu: SelectMenu,
}

impl WebserverScreen {
    pub fn new() -> Self {
        let items = vec![
            "1) Nginx (recommended)".to_string(),
            "2) Apache".to_string(),
            "3) Caddy".to_string(),
        ];
        Self { menu: SelectMenu::new(items) }
    }
}

impl Screen for WebserverScreen {
    fn render(&mut self, area: Rect, frame: &mut Frame, _state: &InstallState) -> crate::error::Result<()> {
        let chunks = Layout::default()
            .direction(Direction::Vertical)
            .margin(1)
            .constraints([
                Constraint::Length(3),  // Title
                Constraint::Length(2),  // Subtitle
                Constraint::Min(10),   // Menu
                Constraint::Length(2),  // Hint
                Constraint::Length(3),  // Footer
            ])
            .split(area);
        
        // Title
        let title = Paragraph::new("Webserver Configuration")
            .style(Style::default().fg(Color::Cyan).add_modifier(Modifier::BOLD))
            .alignment(Alignment::Center);
        frame.render_widget(title, chunks[0]);
        
        // Subtitle
        let subtitle = Paragraph::new("Select your preferred webserver:")
            .style(Style::default().fg(Color::Gray))
            .alignment(Alignment::Center);
        frame.render_widget(subtitle, chunks[1]);
        
        // Menu
        self.menu.render(chunks[2], frame.buffer_mut());
        
        // Hint
        let hint = Paragraph::new("Use ↑↓, 1-3, Enter to select")
            .style(Style::default().fg(Color::Gray))
            .alignment(Alignment::Center);
        frame.render_widget(hint, chunks[3]);
        
        // Footer
        let footer = Paragraph::new("Back: b | Close: c")
            .style(Style::default().fg(Color::Yellow))
            .alignment(Alignment::Center);
        frame.render_widget(footer, chunks[4]);
        
        Ok(())
    }
    
    fn handle_key(&mut self, key: KeyEvent, state: &mut InstallState) -> crate::error::Result<ScreenResult> {
        use crossterm::event::KeyCode;
        
        match key.code {
            KeyCode::Up | KeyCode::Char('k') => self.menu.previous(),
            KeyCode::Down | KeyCode::Char('j') => self.menu.next(),
            KeyCode::Enter => {
                match self.menu.selected_index() {
                    0 => state.webserver = WebserverType::Nginx,
                    1 => state.webserver = WebserverType::Apache,
                    2 => state.webserver = WebserverType::Caddy,
                    _ => {}
                }
                return Ok(ScreenResult::Next(Box::new(super::protocol::ProtocolScreen::new())));
            }
            KeyCode::Char('b') => return Ok(ScreenResult::Back),
            KeyCode::Char('c') => return Ok(ScreenResult::Exit),
            KeyCode::Char(c) if c.is_ascii_digit() => {
                let idx = c.to_digit(10).unwrap() as usize;
                if idx > 0 && idx <= 3 {
                    self.menu.select_by_index(idx - 1);
                }
            }
            _ => {}
        }
        Ok(ScreenResult::None)
    }
}
